### PR TITLE
Fix race condition when multiple gunicorn workers try to download same models

### PR DESCRIPTION
**Pull request by yuruofeifei**
Fix race condition when multiple gunicorn workers try to download same models. #300 
The fix is adding lock between processes to make sure only one worker is downloading and extracting.

Tested on a docker container, no error messages occurs and serving is successful.

@jesterhazy @lupesko

### CODE REVIEW

File: mms/model_loader.py [hunk 1]
```diff
@@ -26,7 +27,20 @@
 MANIFEST_DIR = "manifest_schema"
 MANIFEST_SCHEMA_FILE = 'manifest-schema.json'
 MANIFEST_FILENAME = 'MANIFEST.json'
+LOCK_FILE = '/tmp/tmp_lock_file'
 
+@fasteners.interprocess_locked(LOCK_FILE)
```
Comment (jesterhazy) [inline mms/model_loader.py]: we should have one lock around complete download/extract process

File: mms/model_loader.py [hunk 2]
```diff
@@ -26,7 +27,20 @@
...
+    if not os.path.isdir(model_dir):
+        os.mkdir(model_dir)
+    try:
+        if '.model' in model_file:
+            _extract_zip(model_file, model_dir)
+    except Exception as e:
+        raise Exception('Failed to open model file %s for model %s. Stacktrace: %s'
+                        % (model_file, model_file_prefix , e))
+
+@fasteners.interprocess_locked(LOCK_FILE)
```
Comment (jesterhazy) [inline mms/model_loader.py]: we are still going to download multiple times

File: mms/model_loader.py [hunk 3]
```diff
@@ -26,7 +27,20 @@
...
 MANIFEST_SCHEMA_FILE = 'manifest-schema.json'
 MANIFEST_FILENAME = 'MANIFEST.json'
+LOCK_FILE = '/tmp/tmp_lock_file'
 
+@fasteners.interprocess_locked(LOCK_FILE)
+def _extract_model_archive(model_dir, model_file):
+    if not os.path.isdir(model_dir):
+        os.mkdir(model_dir)
+    try:
+        if '.model' in model_file:
```
Comment (jesterhazy) [inline mms/model_loader.py]: this check should happen before this method is even called. especially now that getting here requires lock acquisition

File: mms/model_loader.py [hunk 4]
```diff
@@ -103,14 +117,8 @@ def _extract_model(service_name, path):
...
 
-    if not os.path.isdir(model_dir):
-        os.mkdir(model_dir)
-    try:
-        if '.model' in model_file:
-            _extract_zip(model_file, model_dir)
-    except Exception as e:
-        raise Exception('Failed to open model file %s for model %s. Stacktrace: %s'
-                        % (model_file, model_file_prefix , e))
+    _extract_model_archive(model_dir, model_file)
```
Comment (jesterhazy) [inline mms/model_loader.py]: single lock boundary should span everything from download to extract, inclusive

File: mms/model_loader.py [hunk 5]
```diff
@@ -45,28 +48,45 @@ def download(url, path=None, overwrite=False):
...
+    model_file = model_location
+    if model_location.lower().startswith(URL_PREFIX):
+        if path is None:
+            model_file = model_location.split('/')[-1]
+        elif os.path.isdir(path):
+            model_file = os.path.join(path, model_location.split('/')[-1])
+        else:
+            model_file = path
+
+        if overwrite or not os.path.exists(model_file):
```
Comment (jesterhazy) [inline mms/model_loader.py]: each worker process will download and extract model to same location, duplicating work.

the lock will serialize access to this function, so results will be correct, but startup time will be multiplied by # of workers.

Comment (yuruofeifei) [reply mms/model_loader.py]: Yes, model will be extracted and download multiple times. But it should be according to the overwrite parameter. If we change the default behavior to always check existence, we are breaking the flow.

Comment (jesterhazy) [reply mms/model_loader.py]: ok we will fix that separately

**Pull request merged**